packages/protocol/src/client-server-protocol/jsonrpc/base-jsonrpc-glsp-client.spec.ts

```diff
@@ -0,0 +1,220 @@
+/********************************************************************************
+ * Copyright (c) 2023 EclipseSource and others.
+ *
+ * This program and the accompanying materials are made available under the
+ * terms of the Eclipse Public License v. 2.0 which is available at
+ * http://www.eclipse.org/legal/epl-2.0.
+ *
+ * This Source Code may also be made available under the following Secondary
+ * Licenses when the conditions for such availability set forth in the Eclipse
+ * Public License v. 2.0 are satisfied: GNU General Public License, version 2
+ * with the GNU Classpath Exception which is available at
+ * https://www.gnu.org/software/classpath/license.html.
+ *
+ * SPDX-License-Identifier: EPL-2.0 OR GPL-2.0 WITH Classpath-exception-2.0
+ ********************************************************************************/
+/* eslint-disable @typescript-eslint/no-empty-function */
+import { expect } from 'chai';
+import * as sinon from 'sinon';
+import { Disposable, Event, MessageConnection, NotificationHandler, ProgressType } from 'vscode-jsonrpc';
+import { remove } from '../../utils/array-util';
+import { expectToThrowAsync } from '../../utils/test-util.spec';
+import { ClientState } from '../glsp-client';
+import { BaseJsonrpcGLSPClient } from './base-jsonrpc-glsp-client';
+import { JsonrpcGLSPClient } from './glsp-jsonrpc-client';
+
+class StubMessageConnection implements MessageConnection {
+    private mockEvent: Event<any> = (listener: (e: any) => any, thisArgs?: any, disposables?: Disposable[]): Disposable =>
+        Disposable.create(() => {});
+
+    sendRequest(...args: any[]): any {
+        throw new Error('Method not implemented.');
+    }
+
+    onRequest(...args: any[]): Disposable {
+        return Disposable.create(() => {});
+    }
+    hasPendingResponse(): boolean {
+        return false;
+    }
+    sendNotification(...args: any[]): Promise<void> {
+        return Promise.resolve();
+    }
+
+    onNotification(...args: any[]): Disposable {
+        return Disposable.create(() => {});
+    }
+
+    onProgress<P>(type: ProgressType<P>, token: string | number, handler: NotificationHandler<P>): Disposable {
+        throw new Error('Method not implemented.');
+    }
+    sendProgress<P>(type: ProgressType<P>, token: string | number, value: P): Promise<void> {
+        throw new Error('Method not implemented.');
+    }
+    onUnhandledProgress = this.mockEvent;
+
+    trace(...args: any[]): Promise<void> {
+        return Promise.resolve();
+    }
+    onError = this.mockEvent;
+    onClose = this.mockEvent;
+    listen(): void {}
+    onUnhandledNotification = this.mockEvent;
+    end(): void {}
+    onDispose = this.mockEvent;
+    dispose(): void {}
+    inspect(): void {}
+}
+
+describe('Base JSON-RPC GLSP Client', () => {
```

Comment:
Very nice test cases for the base GLSP client implementation! If I could only have one request, it would be that it also tests the `onActionMessage` method since this is part of the public API. Not sure how easy it is with the mock implementation though.

Reply:
We could add a test to verify that the corresponding method of the underlying message connection is triggered (`connection.onNotification(JsonrpcGLSPClient.ActionMessageNotification, handler)`) but other than that i don't think additional test cases makes sense because we would either test a artificial mock implementation that does not have anything in common with the actual handling or we would test the functionality provided by vscode-jsonrpc.